File: app/src/main/java/com/esbati/keivan/persiancalendar/Features/Notification/NotificationHelper.java
package com.esbati.keivan.persiancalendar.Features.Notification;

import android.app.AlarmManager;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.support.v7.app.NotificationCompat;
import android.text.TextUtils;

import com.esbati.keivan.persiancalendar.Components.ApplicationController;
import com.esbati.keivan.persiancalendar.Features.Home.MainActivity;
import com.esbati.keivan.persiancalendar.POJOs.CalendarDay;
import com.esbati.keivan.persiancalendar.R;
import com.esbati.keivan.persiancalendar.Repository.PreferencesHelper;
import com.esbati.keivan.persiancalendar.Utils.ColorHelper;
import com.esbati.keivan.persiancalendar.Utils.LanguageHelper;

import java.util.Calendar;

/**
 * Created by Keivan Esbati on 4/9/2017.
 */

public class NotificationHelper {
    private static final int STICKY_NOTIFICATION_ID = 666;
    private static final int[] NOTIFICATION_PRIORITY = {
            NotificationCompat.PRIORITY_MIN
            , NotificationCompat.PRIORITY_LOW
            , NotificationCompat.PRIORITY_DEFAULT
            , NotificationCompat.PRIORITY_HIGH
            , NotificationCompat.PRIORITY_MAX
    };

    public static void showStickyNotification(CalendarDay shownDay){
        Context context = ApplicationController.getContext();

        //Setup Content Intent
        Intent intent = new Intent(context, MainActivity.class);
        int requestId = (int)System.currentTimeMillis(); //unique requestID to differentiate between various notification with same Id
        int flags = PendingIntent.FLAG_CANCEL_CURRENT; // cancel old intent and create new one
        PendingIntent pIntent = PendingIntent.getActivity(context, requestId, intent, flags);


        //Setup Notification
        int notificationPriority = NOTIFICATION_PRIORITY[PreferencesHelper.loadInt(PreferencesHelper.KEY_NOTIFICATION_PRIORITY, 2)];
        NotificationCompat.Builder mBuilder = (NotificationCompat.Builder) new NotificationCompat.Builder(context)
                .setPriority(notificationPriority)
                .setColor(ColorHelper.getSeasonColor(shownDay.mPersianDate.getPersianMonth()))
                .setSmallIcon(R.drawable.icon01 + shownDay.mDayNo - 1)
                .setContentIntent(pIntent)
                .setWhen(0)
                .setShowWhen(false)
                .setAutoCancel(false)
                .setOngoing(true);

        //Setup Title Text
        String mTitle = LanguageHelper.formatStringInPersian(shownDay.mPersianDate.getPersianLongDate());
        mBuilder.setContentTitle(mTitle);

        //Set Content Text
        String collapsedText = "";
        if(shownDay.mGoogleEvents != null)
            //Find an Event with Title
            for(int i = 0 ; i < shownDay.mGoogleEvents.size() ; i++)
                if(!TextUtils.isEmpty(shownDay.mGoogleEvents.get(i).mTITLE)){
                    collapsedText += shownDay.mGoogleEvents.get(i).mTITLE ;
                    break;
                }

        //Adjust Content Text
        if(!TextUtils.isEmpty(collapsedText)){
            //If an Event with Title is Found and R est of Events Count
            if(shownDay.mGoogleEvents.size() > 1)
                collapsedText += " و " + (shownDay.mGoogleEvents.size() - 1) + " رویداد دیگر";
        } else {
            //If No Event with Title is Found just Add Events Count
            if(shownDay.mGoogleEvents.size() > 0)
                collapsedText = shownDay.mGoogleEvents.size() + " رویداد";
        }
        mBuilder.setContentText(collapsedText.trim());

        //If Events are more than One Create Expanded Inbox Style View
        if(shownDay.mGoogleEvents != null && shownDay.mGoogleEvents.size() > 1){
            NotificationCompat.InboxStyle inboxStyle = new NotificationCompat.InboxStyle();
            String[] events = new String[shownDay.mGoogleEvents.size()];

            for(int i = 0 ; i < shownDay.mGoogleEvents.size() ; i++)
                events[i] = !TextUtils.isEmpty(shownDay.mGoogleEvents.get(i).mTITLE) ? shownDay.mGoogleEvents.get(i).mTITLE : shownDay.mGoogleEvents.get(i).mDESCRIPTION;

            inboxStyle.setBigContentTitle(shownDay.mPersianDate.getPersianLongDate());
            for (int i = 0 ; i < events.length ; i++)
                inboxStyle.addLine(events[i]);

            mBuilder.setStyle(inboxStyle);
        }

        //Setup Actions
        if(PreferencesHelper.isOptionActive(PreferencesHelper.KEY_NOTIFICATION_ACTIONS, true)){
            mBuilder.addAction(R.drawable.ic_server_remove_white_24dp, "Dismiss", NotificationActionService.Companion.getDismissAction(context));
        }

        //Show Notification
        NotificationManager mNotificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        mNotificationManager.notify(STICKY_NOTIFICATION_ID, mBuilder.build());

        //Register Alarm to Trigger in case of Broadcast Failed and Service Killed
        registerAlarm(context);
        enableReceiver(context);
    }

    public static void cancelNotification(){
        ((NotificationManager) ApplicationController
                .getContext()
                .getSystemService(Context.NOTIFICATION_SERVICE))
                .cancel(STICKY_NOTIFICATION_ID);

    }

    private static void registerAlarm(Context context){
        AlarmManager alarmManager = (AlarmManager)context.getSystemService(Context.ALARM_SERVICE);
        Intent intent = new Intent(context, NotificationBroadcastReceiver.class);
        PendingIntent alarmIntent = PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);

        // Set the alarm to start at midnight
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        calendar.set(Calendar.HOUR_OF_DAY, 0);
        calendar.set(Calendar.MINUTE, 0);

        alarmManager.setRepeating(AlarmManager.RTC_WAKEUP
                , calendar.getTimeInMillis()
                , AlarmManager.INTERVAL_DAY
                , alarmIntent);
    }

    private static void enableReceiver(Context context) {
        ComponentName receiver = new ComponentName(context, NotificationBroadcastReceiver.class);
        PackageManager pm = context.getPackageManager();

        pm.setComponentEnabledSetting(receiver,
                PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
                PackageManager.DONT_KILL_APP);
    }

    private static void disableReceiver(Context context) {
        ComponentName receiver = new ComponentName(context, NotificationBroadcastReceiver.class);
        PackageManager pm = context.getPackageManager();

        pm.setComponentEnabledSetting(receiver,
                PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
                PackageManager.DONT_KILL_APP);
    }
}

File: app/src/main/java/com/esbati/keivan/persiancalendar/features/notification/NotificationHelper.kt
package com.esbati.keivan.persiancalendar.features.notification

import android.annotation.TargetApi
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import android.provider.Settings
import androidx.core.app.NotificationCompat
import android.text.TextUtils
import com.esbati.keivan.persiancalendar.BuildConfig
import com.esbati.keivan.persiancalendar.R
import com.esbati.keivan.persiancalendar.features.home.MainActivity
import com.esbati.keivan.persiancalendar.pojos.CalendarDay
import com.esbati.keivan.persiancalendar.repository.PreferencesHelper
import com.esbati.keivan.persiancalendar.utils.ColorHelper
import com.esbati.keivan.persiancalendar.utils.LanguageHelper

/**
 * Created by Keivan Esbati on 4/9/2017.
 */

object NotificationHelper {
    const val STICKY_NOTIFICATION_ID = 16
    private const val STICKY_NOTIFICATION_CHANNEL_ID = "STICKY_NOTIFICATION_CHANNEL"
    private val NOTIFICATION_PRIORITY = intArrayOf(
            NotificationCompat.PRIORITY_MIN,
            NotificationCompat.PRIORITY_LOW,
            NotificationCompat.PRIORITY_DEFAULT,
            NotificationCompat.PRIORITY_HIGH,
            NotificationCompat.PRIORITY_MAX
    )

    fun createNotificationChannelIfRequired(context: Context) {
        // Create the NotificationChannel, but only on API 26+ because
        // the NotificationChannel class is new and not in the support library
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val name = context.getString(R.string.notification_sticky_channel_name)
            val descriptionText = context.getString(R.string.notification_sticky_channel_description)
            val importance = NotificationManager.IMPORTANCE_LOW
            val channel = NotificationChannel(STICKY_NOTIFICATION_CHANNEL_ID, name, importance).apply {
                description = descriptionText
                setShowBadge(false)
            }
            // Register the channel with the system
            val notificationManager: NotificationManager =
                    context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.createNotificationChannel(channel)
        }
    }

    @JvmStatic
    @TargetApi(26)
    fun getChannelImportance(context: Context): Int {
        val notificationManager: NotificationManager =
                context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        return notificationManager.getNotificationChannel(STICKY_NOTIFICATION_CHANNEL_ID).importance
    }

    @JvmStatic
    @TargetApi(26)
    fun openChannelSetting(context: Context) {
        val intent = Intent(Settings.ACTION_CHANNEL_NOTIFICATION_SETTINGS).apply {
            putExtra(Settings.EXTRA_APP_PACKAGE, BuildConfig.APPLICATION_ID)
            putExtra(Settings.EXTRA_CHANNEL_ID, NotificationHelper.STICKY_NOTIFICATION_CHANNEL_ID)
        }
        context.startActivity(intent)
    }

    fun showStickyNotification(context: Context, shownDay: CalendarDay) {
        val notification = createStickyNotification(context, shownDay)

        //Show Notification
        (context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager)
                .notify(STICKY_NOTIFICATION_ID, notification )

        NotificationService.startService(context)
    }

    fun createStickyNotification(context: Context, shownDay: CalendarDay): Notification {
        //Setup Content Intent
        val intent = Intent(context, MainActivity::class.java)
        val requestId = System.currentTimeMillis().toInt() //unique requestID to differentiate between various notification with same Id
        val pIntent = PendingIntent.getActivity(context, requestId, intent, PendingIntent.FLAG_CANCEL_CURRENT)

        //Setup Notification
        val notificationPriority = NOTIFICATION_PRIORITY[PreferencesHelper.notificationPriority]
        val mBuilder = NotificationCompat.Builder(context, STICKY_NOTIFICATION_CHANNEL_ID)
                .setPriority(notificationPriority)
                .setColor(ColorHelper.getSeasonColor(shownDay.mMonth, context))
                .setSmallIcon(R.drawable.icon01 + shownDay.mDay - 1)
                .setContentIntent(pIntent)
                .setWhen(0)
                .setShowWhen(false)
                .setAutoCancel(false)
                .setOngoing(true) as NotificationCompat.Builder

        //Setup Title Text
        mBuilder.setContentTitle(LanguageHelper.formatStringInPersian(shownDay.formattedDate))

        //Set Content Text
        if(shownDay.getEventsSummary(context).isNotBlank())
            mBuilder.setContentText(shownDay.getEventsSummary(context))

        //If more than one event is available add an expanded Inbox Style view
        if (shownDay.mEvents.size > 1) {
            val inboxStyle = NotificationCompat.InboxStyle()
            inboxStyle.setBigContentTitle(shownDay.formattedDate)

            for (event in shownDay.mEvents)
                inboxStyle.addLine(
                        if (!TextUtils.isEmpty(event.title))
                            event.title
                        else
                            event.description
                )

            mBuilder.setStyle(inboxStyle)
        }

        //Setup Actions
        if (PreferencesHelper.isNotificationActionsActive) {
            mBuilder.addAction(
                    R.drawable.ic_server_remove_white_24dp
                    , context.getString(R.string.notification_action_dismiss_title)
                    , NotificationActionService.getDismissAction(context)
            )
        }

        return mBuilder.build()
    }

    fun cancelNotification(context: Context) {
        NotificationService.stopService(context)
        (context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager)
                .cancel(STICKY_NOTIFICATION_ID)
    }
}
